components/disk/component_output.go

```diff
@@ -33,90 +34,114 @@ const (
+
+	StateNameDiskExtPartitionsTotal         = "disk_ext_partitions_total"
+	StateKeyDiskExtPartitionsTotalBytes     = "disk_ext_partitions_total_bytes"
+	StateKeyDiskExtPartitionsTotalGB        = "disk_ext_partitions_total_gb"
+	StateKeyDiskExtPartitionsTotalHumanized = "disk_ext_partitions_total_humanized"
 )
 
 func (o *Output) States() ([]components.State, error) {
-	diskExtPartitionsData, err := o.DiskExtPartitions.JSON()
-	if err != nil {
-		return nil, err
+	querySucceededState := components.State{
+		Name:    disk_id.Name,
+		Healthy: true,
+		Reason:  "query succeeded",
 	}
-	diskBlockDevicesData, err := o.DiskBlockDevices.JSON()
-	if err != nil {
-		return nil, err
+
+	var (
+		diskBlockDevicesData      []byte
+		blkDevTotalBytes          uint64
+		blkDevTotalGB             float64
+		blkDevTotalBytesHumanized string
+	)
+	if len(o.DiskBlockDevices) > 0 {
+		var err error
+		diskBlockDevicesData, err = o.DiskBlockDevices.JSON()
+		if err != nil {
+			return nil, err
+		}
+		blkDevTotalBytes = o.DiskBlockDevices.GetTotalBytes()
+		blkDevTotalGB = float64(blkDevTotalBytes) / 1e9
+		blkDevTotalBytesHumanized = humanize.Bytes(blkDevTotalBytes)
 	}
 
-	totalMountedBytes := o.DiskExtPartitions.GetMountedTotalBytes()
-	totalMountedGB := float64(totalMountedBytes) / 1e9
-	totalMountedBytesHumanized := humanize.Bytes(totalMountedBytes)
+	var (
+		diskExtPartitionsData      []byte
+		totalMountedBytes          uint64
+		totalMountedGB             float64
+		totalMountedBytesHumanized string
+	)
+	if len(o.DiskExtPartitions) > 0 {
+		var err error
+		diskExtPartitionsData, err = o.DiskExtPartitions.JSON()
+		if err != nil {
+			return nil, err
+		}
 
-	blkDevTotalBytes := o.DiskBlockDevices.GetTotalBytes()
-	blkDevTotalGB := float64(blkDevTotalBytes) / 1e9
-	blkDevTotalBytesHumanized := humanize.Bytes(blkDevTotalBytes)
+		totalMountedBytes = o.DiskExtPartitions.GetMountedTotalBytes()
+		totalMountedGB = float64(totalMountedBytes) / 1e9
+		totalMountedBytesHumanized = humanize.Bytes(totalMountedBytes)
+	}
 
-	mb, err := json.Marshal(o.MountTargetUsages)
-	if err != nil {
-		return nil, err
+	var mountTargetUsagesBytes []byte
+	if len(o.MountTargetUsages) > 0 {
+		var err error
+		mountTargetUsagesBytes, err = json.Marshal(o.MountTargetUsages)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	return []components.State{
+		querySucceededState,
 		{
-			Name:    disk_id.Name,
-			Healthy: true,
-			Reason:  "query succeeded",
+			Name:    StateNameDiskBlockDevices,
+			Healthy: len(o.DiskBlockDevices) > 0,
+			Reason:  fmt.Sprintf("found %d block devices", len(o.DiskBlockDevices)),
+			ExtraInfo: map[string]string{
+				StateKeyData:     string(diskBlockDevicesData),
+				StateKeyEncoding: StateValueEncodingJSON,
+			},
 		},
 		{
-			Name:    StateNameDiskExtPartition,
-			Healthy: true,
-			Reason:  "",
+			Name:    StateNameDiskBlockDevicesTotal,
+			Healthy: len(o.DiskBlockDevices) > 0,
```

Comment:
maybe just return true for all the disk states for now..noticed findmnt also flaky..

Reply:
Setting true where we use `MountTargetsToTrackUsage` (using findmnt)